### PR TITLE
Backpropagation

### PR DESCRIPTION
**Pull request by molbazsa**

### CODE REVIEW

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            Debug.Assert(weights.RowCount == biases.Count);
+            Weights = weights;
+            Biases = biases;
+            ActivationFunction = activationFunction;
+        }
+
+        public Vector<double> WeightedSums(Vector<double> previousLayer)
+        {
+            Debug.Assert(previousLayer.Count == Weights.ColumnCount);
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs]: Dimensions

- previousLayer (n, 1), where **n** is the number of neuron in the previous layer.
- Weights (n, k), where **k** is the number of neuron in the next layer.
- Bias (k, 1)

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            return Weights * previousLayer + Biases;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs]: Formula: DotProduct(Weights.T, previousLayer) + Biases
Result dimension: (k,n)x(n,1) + (k,1) = (k,1)

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            _Layers = layers;
+        }
+
+        public Vector<double> PropagateForward(Vector<double> inputData)
+        {
+            Vector<double> activations = inputData;
+
+            foreach (Layer layer in _Layers)
+            {
+                activations = layer.Activations(activations);
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs]: A = inputData
for i .. layer
   weightedSum = Layer.WeightedSums(A)
   A = Layer.Activation(weightedSum)

**Review by SzilvasiPeter — CHANGES_REQUESTED** [review]

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs (hunk 1 above)
Comment (molbazsa) [reply LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs]: These conditions already hold true.

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs (hunk 2 above)
Comment (molbazsa) [reply LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs]: The `Matrix<> * Vector<>` operator provides the correct functionality.

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs (hunk 3 above)
Comment (molbazsa) [reply LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs]: The calculation is fixed by the upcoming commit.

**Comment by molbazsa** [comment]
I got stuck implementing the rest of the backpropagation algorithm, so I think I will play around with a one neuron per layer case before moving on to this complete implementation.
@SzilvasiPeter I think we can merge this branch in this state, because it doesn't break anything so far. That way the main branch can be brought to a stable state, and I don't want to hold up the process as the implementation might take some time.

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs [hunk 4]
```diff
@@ -25,18 +30,29 @@ public Layer(Matrix<double> weights, Vector<double> biases, IActivationFunction
...
         {
             return Vector<double>.Build.Dense(
                 weightedSums.Count,
                 i => ActivationFunction.Activation(weightedSums[i])
             );
         }
+
+        public void Update(double rateOfChange, double learningRate)
+        {
+            Weights += learningRate * rateOfChange * WeightsGradient;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs]: The correct formula would be:
weights = weights - alpha*weightsGradient

Same for the bias:
bias = bias - alpha*biasGradient

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs [hunk 5]
```diff
@@ -21,48 +21,55 @@ public Vector<double> PropagateForward(Vector<double> inputData)
...
+            Vector<double> cumulativeDerivatives = CostDerivatives(actualValues, expectedValues);
 
             throw new NotImplementedException();
         }
 
-        private Vector<double> ActivationDerivatives(Vector<double> actualValue, Vector<double> expectedValue)
+        private Vector<double> CostDerivatives(Matrix<double> actualValues, Matrix<double> expectedValues)
+        {
+            Vector<double> derivatives = Vector<double>.Build.Dense(actualValues.RowCount, 1);
+            derivatives = (actualValues - expectedValues) * derivatives;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs:53]: `derivatives` will be (n,1) vector, containing only 1s. Why do we need to multiply by a vector only containing 1s?

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs (hunk 4 above)
Comment (molbazsa) [reply LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Layer.cs]: Yeah, I messed up the sign. I included the `rateOfChange` to make the step in the negative gradient direction proportional to the partial derivative at that point. I saw this in the 3blue1brown videos and thought we could use it.

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs (hunk 5 above)
Comment (molbazsa) [reply LinearRegression/LinearRegressionBackend/MLNeuralNetwork/NeuralNetwork.cs:53]: The multiplication is for summing up the derivatives in each column of the matrix, where each column corresponds to a piece of testing data. I thought maybe this is faster than doing it with a for loop.

**Review by SzilvasiPeter — APPROVED** [review]

**Comment by SzilvasiPeter** [comment]
> I got stuck implementing the rest of the backpropagation algorithm, so I think I will play around with a one neuron per layer case before moving on to this complete implementation. @SzilvasiPeter I think we can merge this branch in this state, because it doesn't break anything so far. That way the main branch can be brought to a stable state, and I don't want to hold up the process as the implementation might take some time.

All right, I will merge the pull request.

**Pull request merged**